### DISCUSSION TITLE
how does Avalonia turn XAML into code?

### DISCUSSION THREAD

**Discussion by BenMakesGames** [AvaloniaUI/Avalonia#6414] [Q&A] [answered]
I've been playing with C# Source Generators recently, and I'm impressed with the idea of turning text - like XML - into code. I also know that Source Generators are very new to C# - introduced in 9.0 - so I assume Avalonia isn't using that.

what does Avalonia use to turn XAML into code? (I'm pretty sure it isn't interpreted at run-time, since there are compile-time errors if the XAML is incorrect.)

**Comment by jmacato** [comment, marked as answer]
We use [XamlX](https://github.com/kekekeks/XamlX) XAML to IL code converter designed and maintained by our very own @kekekeks . It translates Avalonia XAML into code so that we do fancy stuff like compiler optimization and compile-time errors as well. We used to translate XAML in runtime with XML parsers but as experience goes it was horribly inefficient :)

**Reply by maxkatz6** [reply]
Worth to mention, XamlX works with any .NET language, not just with C#.

**Reply by jmacato** [reply]
yes, that too. It doesnt compile to C# but rather directly to what the .NET runtime uses (MSIL).
